service/matching/task_validation.go

```diff
@@ -0,0 +1,195 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+
+	"go.temporal.io/server/api/historyservice/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/primitives/timestamp"
+)
+
+const (
+	taskReaderOfferTimeout        = 60 * time.Second
+	taskReaderValidationThreshold = 600 * time.Second
+)
+
+type (
+	taskValidator interface {
+		maybeValidate(
+			task *persistencespb.AllocatedTaskInfo,
+			taskType enumspb.TaskQueueType,
+		) bool
+	}
+
+	taskValidationInfo struct {
+		taskID         int64
+		validationTime time.Time
+	}
+
+	taskValidatorImpl struct {
+		historyClient historyservice.HistoryServiceClient
+
+		lastValidatedTaskInfo taskValidationInfo
+	}
+)
+
+func newTaskValidator(
+	historyClient historyservice.HistoryServiceClient,
+) *taskValidatorImpl {
+	return &taskValidatorImpl{
+		historyClient: historyClient,
+	}
+}
+
+// check if a task has expired / is valid
+// if return false, then task is invalid and should be discarded
+// if return true, then task is *not invalid*, and should be dispatched
```

Comment:
"maybe-valid" feels easier to read?
```suggestion
// if return true, then task is *maybe-valid*, and should be dispatched
```